`mul`: port to structured

bdhirsh opened the pull request:
Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54085 fix torch.pow type promotion issue
* #53680 port div to structured
* #53679 port sub to structured, fix sub.Scalar bug
* #53669 port at::pow to structured, fix pow bug with mixed cpu/cuda tensors
* **#52692 port at::mul to structured**

Differential Revision: [D27029822](https://our.internmc.facebook.com/intern/diff/D27029822)

bdhirsh committed:
port at::mul to structured

[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"




[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"




[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -126,7 +127,7 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
       return result.zero_();
     }
-    return at::native::mul_out(result, self, at::native::scalar_tensor(beta, at::device(at::kCPU).dtype(self.scalar_type())));
+    return at::cuda::mul_out(result, self, at::native::scalar_tensor(beta, at::device(at::kCPU).dtype(self.scalar_type())));
```
Technically, I think `at::cpu::mul_out` is more in-line with the original behavior than `at::cuda::mul_out`. They both eventually call the structured kernel implementation, which uses DispatchStub to dispatch appropriately to cpu/cuda.

The only difference is that `at::cuda::mul_out` sets up a cuda guard before calling into the kernel. At this point in the code we're already inside a cuda kernel, so I think that the extra guard setup might be redundant.

To me it seems more in the right spirit to call the cuda variant than the cpu variant, since:
- Calling the cpu variant here would look confusing to people who don't know what `at::cpu::mul_out` actually does
- calling the cuda variant protects against any future divergences in the cuda/cpu implementations.

ezyang replied:
Clearly this means we should have a `cuda_unchecked` namespace lololol

ezyang replied:
(Switching this to cuda is fine by me)

ezyang replied:
@robieta It suuuuure would be nice if we had a comprehensive set of microbenchmarks that would notice when an extra device guard was called, as it is in this case :)

bdhirsh commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -391,7 +391,7 @@ void Reducer::copy_grad_to_bucket(
     wrapped.unsafeGetTensorImpl()->set_wrapped_number(true);
     // Divides while copying into the bucket view.
-    at::native::mul_out(bucket_view, grad, wrapped);
+    at::mul_out(bucket_view, grad, wrapped);
```
So this was weird: I originally updated this to call `at::cpu::mul_out`, and I #included the `ATen/CPUFunctions.h` at the top of the file. That all compiled and linked without any errors, but when I tried to import torch from python I got some weird undefined symbols:

```
>>> import torch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/raid/hirsheybar/pytorch-port_structured_TI/torch/__init__.py", line 197, in <module>
    from torch._C import *
ImportError: /raid/hirsheybar/pytorch-port_structured_TI/torch/lib/libtorch_python.so: undefined symbol: _ZN2at3cpu7mul_outERNS_6TensorERKS1_S4_
```

I ran `c++filt` on that symbol to de-mangle it:
```
> c++filt _ZN2at3cpu7mul_outERNS_6TensorERKS1_S4_
at::cpu::mul_out(at::Tensor&, at::Tensor const&, at::Tensor const&)
```

That confirmed that it was coming from from my usage of `at::cpu::mul_out`, which I pinpointed to be from this file. I'm still not sure why that gave me an error, and it wasn't caught by the linker. The function implementation lives in `RegisterCPU.cpp`, and it's possible that the build workflow for `c10d` doesn't build that file or something.

Either way, I think it's probably fine to use `at::mul_out` here instead. It'll require another trip through the dispatcher, but that seems like less of a performance issue for c10d since it's distributed.

ezyang replied:
This suggests c10d links against libtorch.so but not libtorch_cpu.so. Hmm, I wonder if we should have a TORCH_CPU_API macro. This workaround seems fine.

bdhirsh committed:
Update on "port at::mul to structured"

Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.




[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"

Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.




[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"

Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.




[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"

Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.




[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"

Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.




[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"

Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.


Differential Revision: [D27029822](https://our.internmc.facebook.com/intern/diff/D27029822)

[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"

Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.


Differential Revision: [D27029822](https://our.internmc.facebook.com/intern/diff/D27029822)

[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"

Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.


Differential Revision: [D27029822](https://our.internmc.facebook.com/intern/diff/D27029822)

[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"

Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.


Differential Revision: [D27029822](https://our.internmc.facebook.com/intern/diff/D27029822)

[ghstack-poisoned]

bdhirsh committed:
Update on "port at::mul to structured"

Porting `at::mul` to structured.

One other issue I hit with the port was the fact that there are a bunch of other places around the code base that used to call out to variants of `at::native::mul`, which no longer exists. *Technically*, `at::cpu::mul` does the equivalent thing now, so I patched most call-sites to use that. There were two other places where I did something slightly different (calling `at::cuda::mul` and `at::mul`, respectively), which I called out in the comments.


Differential Revision: [D27029822](https://our.internmc.facebook.com/intern/diff/D27029822)

[ghstack-poisoned]
